src/OpenTelemetry/Metrics/CompositeMetricReader.cs

```diff
@@ -57,12 +60,117 @@ public CompositeMetricReader AddReader(MetricReader reader)
+
+            int index = 0;
+            for (var cur = this.head; cur != null; cur = cur.Next)
+            {
+                if (metrics[index] != null)
+                {
+                    cur.Value.RecordSingleStreamLongMeasurement(metrics[index], value, tags);
+                }
+
+                index++;
+            }
+        }
+
+        internal void RecordSingleStreamDoubleMeasurements(List<Metric> metrics, double value, ReadOnlySpan<KeyValuePair<string, object>> tags)
+        {
+            Debug.Assert(metrics.Count == this.count, "The count of metrics to be updated for a CompositeReader must match the number of individual readers.");
+
+            int index = 0;
+            for (var cur = this.head; cur != null; cur = cur.Next)
+            {
+                if (metrics[index] != null)
+                {
+                    cur.Value.RecordSingleStreamDoubleMeasurement(metrics[index], value, tags);
+                }
+
+                index++;
+            }
+        }
+
+        internal List<List<Metric>> AddMetricsSuperListWithViews(Instrument instrument, List<MetricStreamConfiguration> metricStreamConfigs)
+        {
+            var metricsSuperList = new List<List<Metric>>();
+            for (var cur = this.head; cur != null; cur = cur.Next)
+            {
+                var metrics = cur.Value.AddMetricsListWithViews(instrument, metricStreamConfigs);
+                metricsSuperList.Add(metrics);
+            }
+
+            return metricsSuperList;
+        }
+
+        internal void RecordLongMeasurements(List<List<Metric>> metricsSuperList, long value, ReadOnlySpan<KeyValuePair<string, object>> tags)
+        {
+            Debug.Assert(metricsSuperList.Count == this.count, "The count of metrics to be updated for a CompositeReader must match the number of individual readers.");
+
+            int index = 0;
+            for (var cur = this.head; cur != null; cur = cur.Next)
+            {
+                if (metricsSuperList[index].Count > 0)
+                {
+                    cur.Value.RecordLongMeasurement(metricsSuperList[index], value, tags);
+                }
+
+                index++;
+            }
+        }
+
+        internal void RecordDoubleMeasurements(List<List<Metric>> metricsSuperList, double value, ReadOnlySpan<KeyValuePair<string, object>> tags)
+        {
+            Debug.Assert(metricsSuperList.Count == this.count, "The count of metrics to be updated for a CompositeReader must match the number of individual readers.");
+
+            int index = 0;
+            for (var cur = this.head; cur != null; cur = cur.Next)
+            {
+                if (metricsSuperList[index].Count > 0)
+                {
+                    cur.Value.RecordDoubleMeasurement(metricsSuperList[index], value, tags);
+                }
+
+                index++;
+            }
+        }
+
+        internal void CompleteSingleStreamMeasurements(List<Metric> metrics)
```

Comment:
Consider putting these in a separate file, too (and later we might consider moving them to a separate class).